src/data/list/prod_sigma.lean

```diff
@@ -0,0 +1,67 @@
+/-
+Copyright (c) 2014 Parikshit Khanna. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Parikshit Khanna, Jeremy Avigad, Leonardo de Moura, Floris van Doorn, Mario Carneiro
+-/
+import data.list.big_operators
+
+/-!
+# Lists in product and sigma types
+
+This file proves basic properties of `list.product` and `list.sigma`, which are list constructions
+living in product and sigma types respectively. Their definitions can be found in
+[`data.list.defs`](./data/list/defs). Beware, this is not about `list.prod`, the multiplicative
+product.
+-/
+
+variables {α β : Type*}
+
+namespace list
+
+/-! ### product -/
+
+@[simp] lemma nil_product (l : list β) : product (@nil α) l = [] := rfl
+
+@[simp] lemma product_cons (a : α) (l₁ : list α) (l₂ : list β)
+        : product (a::l₁) l₂ = map (λ b, (a, b)) l₂ ++ product l₁ l₂ := rfl
+
+@[simp] lemma product_nil : ∀ (l : list α), product l (@nil β) = []
+| []     := rfl
+| (a::l) := by rw [product_cons, product_nil]; refl
+
+@[simp] lemma mem_product {l₁ : list α} {l₂ : list β} {a : α} {b : β} :
+  (a, b) ∈ product l₁ l₂ ↔ a ∈ l₁ ∧ b ∈ l₂ :=
+by simp only [product, mem_bind, mem_map, prod.ext_iff, exists_prop,
+  and.left_comm, exists_and_distrib_left, exists_eq_left, exists_eq_right]
+
+lemma length_product (l₁ : list α) (l₂ : list β) :
+  length (product l₁ l₂) = length l₁ * length l₂ :=
+by induction l₁ with x l₁ IH; [exact (zero_mul _).symm,
+  simp only [length, product_cons, length_append, IH,
+    right_distrib, one_mul, length_map, add_comm]]
```

Comment:
This seems to be from lean core (https://github.com/leanprover-community/mathlib/commit/deb168147d4ede8848940a797753f026648501a4#diff-493f87cfdba1f4af0572b6abb86b2f0205ef9829abcb70b8453e4e5f8aecc59d), with authors "Leonardo de Moura, Haitao Zhang, Floris van Doorn"

Reply:
Actually, it's just Leo:

https://github.com/leanprover/lean/blob/9a7cff0e89afe0f5e70baa60aaaa37629f60ad42/library/data/list/comb.lean